### ISSUE TITLE
Crash on arm64 page_size 64k

### ISSUE THREAD

**Issue by tosone** [upx/upx#734]
This issue tracker is **ONLY** used for reporting bugs.
Please use [stackoverflow](https://stackoverflow.com) for supporting issues.

## What's the problem (or question)?

Crash on arm64 64k page_size machine. but 4k machine runs well.

``` bash
upx --version
```

## What should have happened?

runs well.

## Do you have an idea for a solution?

After make this change: https://github.com/tosone/upx/commit/7061d333200f8461d4c4f4df8570ef3e400c54ea, everything is ok.

## How can we reproduce the issue?

arm64 64k page_size machine run `upx --version`

## Please tell us details about your environment.

* UPX version used (`upx --version`):
v4.2.1.
* Host Operating System and version:
debian
* Host CPU architecture:
arm64
* Target CPU architecture:
arm64

**Comment by markus-oberhumer** [comment]
I fear I don't get you, is this a bug report about our official arm64_linux release binary, or about your own self-compiled version?

**Comment by tosone** [comment]
official arm64_linux binary crash on my arm64 page_size 64k machine. just show me `segmention fault`

**Comment by tosone** [comment]
https://github.com/tosone/upx/commit/7061d333200f8461d4c4f4df8570ef3e400c54ea After i compile with my patch. It will be ok.

**Comment by markus-oberhumer** [comment]
Does the unpacked official arm64_linux binary work?

**Comment by tosone** [comment]
unpacked? I just downloaded the upx from release page. https://github.com/upx/upx/releases/download/v4.2.1/upx-4.2.1-arm64_linux.tar.xz

![img_v3_0250_762bdaab-1580-4be2-b6f8-a0e7636aa11g](https://github.com/upx/upx/assets/5346506/58f12998-abb1-4ffe-931d-a3ed8ecc34a1)

**Comment by markus-oberhumer** [comment]
Use `some_working_upx_version -d upx -o upx_unpacked`

**Comment by tosone** [comment]
My god. upx_unpacked is ok!!!

**Comment by markus-oberhumer** [comment]
Great, could you also please test our previous official arm64_linux versions, i.e. version 4.0.0 and later, both packed and unpacked? Thanks!

**Comment by tosone** [comment]
4.0.2: packed crash. unpacked ok.
4.1.0: packed crash. unpacked ok.
4.2.0: packed crash. unpacked ok.

**Comment by tosone** [comment]
I use the upx that compiled by myself unpack the official upx to upx_unpacked, upx_unpacked is ok.

then upx_unpacked pack it self, crash.

but the upx that I compile by myself, self packed file is ok. So curious.

script here: 
``` bash
DARCH="arm64"
UPX_VERSION="4.2.1"
wget -O upx https://fs.tosone.cn/vecps-cli/addons/upx/upx-arm64-20231108 && \ # upx compiled by myself
  chmod +x ./upx && mv ./upx /usr/local/bin && upx --version && \ # ok
  wget https://github.com/upx/upx/releases/download/v${UPX_VERSION}/upx-${UPX_VERSION}-${DARCH}_linux.tar.xz && \
  tar xf upx-${UPX_VERSION}-${DARCH}_linux.tar.xz && \
  cd upx-${UPX_VERSION}-${DARCH}_linux && \
  upx -d upx -o upx_unpacked && ./upx_unpacked --version && \ # ok
  upx -4 -o upx_compile_myself_packed /usr/local/bin/upx && ./upx_compile_myself_packed --version && \ # ok
  ./upx_unpacked -4 -o upx_packed ./upx_unpacked && ./upx_packed --version # crash
```

**Comment by jreiser** [comment]
@tosone I'm having trouble finding an easy method to install arm64 Debian system on a RaspberryPi model 4 (or model 3).  Please share any pointers that you may have.

**Comment by tosone** [comment]
I test on the normal arm64 page_size 4k machine, it is ok. But crashes in my company ci pipeline which is arm64 page_size 64k machine.

**Comment by jreiser** [comment]
@tosone Please try the above commit on branch `devel4`.  I lack both software and a machine for 64kB page size.

**Comment by tosone** [comment]
Sorry for the late resp. 

My compile command is `apt install cmake build-essential && make` then get upx from `build/release`.
g++ version is `g++ (Debian 12.2.0-14) 12.2.0`

`upx-20231120` and `upx.mini-20231120` is compiled with the `devel4`.
``` sh
./upx-20231120 -4 -o upx.mini-20231120 ./upx-20231120
```

`upx-v4.2.1` and `upx.mini-v4.2.1` is compiled after `rm -rf build/` and `checkout v4.2.1`.
``` sh
./upx-v4.2.1 -4 -o upx.mini-v4.2.1 ./upx-v4.2.1
```

```
export DARCH="arm64"
export UPX_VERSION="4.2.1"
wget -O upx-20231120 https://fs.tosone.cn/upx-20231120 && \
wget -O upx.mini-20231120 https://fs.tosone.cn/upx.mini-20231120 && \
wget -O upx-v4.2.1 https://fs.tosone.cn/upx-v4.2.1 && \
wget -O upx.mini-v4.2.1 https://fs.tosone.cn/upx.mini-v4.2.1 && \
chmod +x ./upx-20231120 && ./upx-20231120 --version && \
chmod +x ./upx.mini-20231120 && ./upx.mini-20231120 --version && \
chmod +x ./upx-v4.2.1 && ./upx-v4.2.1 --version && \
chmod +x ./upx.mini-v4.2.1 && ./upx.mini-v4.2.1 --version && \
./upx-20231120 -9 -o upx-20231120_packed ./upx-20231120 && ./upx-20231120_packed --version && \
./upx-v4.2.1 -9 -o upx-v4.2.1_packed ./upx-v4.2.1 && ./upx-v4.2.1_packed --version && \
wget https://github.com/upx/upx/releases/download/v${UPX_VERSION}/upx-${UPX_VERSION}-${DARCH}_linux.tar.xz && \
tar xf upx-${UPX_VERSION}-${DARCH}_linux.tar.xz && \
cd upx-${UPX_VERSION}-${DARCH}_linux && \
./upx --version # crash here, all of the above is ok.
```

I cannot reproduce the error compiled by myself. Both `v4.2.1` and `devel4` are ok.

What have we done with the release upx.tar.gz file? or could you please compile a `upx`, and I can try it.

**Comment by jreiser** [comment]
@tosone The auto-build for any Commit is available by clicking on the Actions item near the top of this page, then clicking on the title "Fix varying page size on arm64", then scrolling down that page to see the Assets which are the builds for varying architectures.
upx.tar.gz is for a Release, which has not been made yet.

**Comment by tosone** [comment]
@jreiser  I test this package `upx-devel-d4e94f5-zigcc-aarch64-linux-musl`.

``` sh
wget -O upx-devel-d4e94f5-zigcc-aarch64-linux-musl.zip https://fs.tosone.cn/upx-devel-d4e94f5-zigcc-aarch64-linux-musl.zip && \
unzip upx-devel-d4e94f5-zigcc-aarch64-linux-musl.zip && tar zxvf upx-devel-d4e94f5-zigcc-aarch64-linux-musl.tar.gz && \
cd upx-devel-d4e94f5-zigcc-aarch64-linux-musl/zig/aarch64-linux-musl/release && ./upx --version && \
./upx -9 -o upx_devel4_packed ./upx && ./upx_devel4_packed --version
```

The output log is here:

```
#9 25.36 2023-11-21 05:18:07 (44.6 MB/s) - 'upx-devel-d4e94f5-zigcc-aarch64-linux-musl.zip' saved [9740978/9740978]
#9 25.36 
#9 25.36 Archive:  upx-devel-d4e94f5-zigcc-aarch64-linux-musl.zip
#9 25.36   inflating: upx-devel-d4e94f5-zigcc-aarch64-linux-musl.tar.gz  
#9 25.45 upx-devel-d4e94f5-zigcc-aarch64-linux-musl/
#9 25.45 upx-devel-d4e94f5-zigcc-aarch64-linux-musl/zig/
#9 25.45 upx-devel-d4e94f5-zigcc-aarch64-linux-musl/zig/aarch64-linux-musl/
#9 25.45 upx-devel-d4e94f5-zigcc-aarch64-linux-musl/zig/aarch64-linux-musl/debug/
#9 25.45 upx-devel-d4e94f5-zigcc-aarch64-linux-musl/zig/aarch64-linux-musl/debug/upx
#9 25.61 upx-devel-d4e94f5-zigcc-aarch64-linux-musl/zig/aarch64-linux-musl/release/
#9 25.61 upx-devel-d4e94f5-zigcc-aarch64-linux-musl/zig/aarch64-linux-musl/release/upx
#9 25.76 upx 4.3.0-git-d4e94f508613
#9 25.76 UCL data compression library 1.03
#9 25.76 zlib data compression library 1.3.0.1-motley
#9 25.76 LZMA SDK version 4.43
#9 25.76 doctest C++ testing framework version 2.4.11
#9 25.76 Copyright (C) 1996-2023 Markus Franz Xaver Johannes Oberhumer
#9 25.76 Copyright (C) 1996-2023 Laszlo Molnar
#9 25.76 Copyright (C) 2000-2023 John F. Reiser
#9 25.76 Copyright (C) 2002-2023 Jens Medoch
#9 25.76 Copyright (C) 1995-2023 Jean-loup Gailly and Mark Adler
#9 25.76 Copyright (C) 1999-2006 Igor Pavlov
#9 25.76 Copyright (C) 2016-2023 Viktor Kirilov
#9 25.76 UPX comes with ABSOLUTELY NO WARRANTY; for details type 'upx -L'.
#9 25.76                        Ultimate Packer for eXecutables
#9 25.76                           Copyright (C) 1996 - 2023
#9 33.71 UPX git-d4e94f  Markus Oberhumer, Laszlo Molnar & John Reiser    Nov 2nd 2023
#9 33.71 
#9 33.71         File size         Ratio      Format      Name
#9 33.71    --------------------   ------   -----------   -----------
#9 33.71   17765792 ->   4455404   25.08%   linux/arm64   upx_devel4_packed
#9 33.71 
#9 33.71 Packed 1 file.
#9 33.71 
#9 33.71 WARNING: this is an unstable beta version - use for testing only! Really.
#9 33.73 upx 4.3.0-git-d4e94f508613
#9 33.73 UCL data compression library 1.03
#9 33.73 zlib data compression library 1.3.0.1-motley
#9 33.73 LZMA SDK version 4.43
#9 33.73 doctest C++ testing framework version 2.4.11
#9 33.73 Copyright (C) 1996-2023 Markus Franz Xaver Johannes Oberhumer
#9 33.73 Copyright (C) 1996-2023 Laszlo Molnar
#9 33.73 Copyright (C) 2000-2023 John F. Reiser
#9 33.73 Copyright (C) 2002-2023 Jens Medoch
#9 33.73 Copyright (C) 1995-2023 Jean-loup Gailly and Mark Adler
#9 33.73 Copyright (C) 1999-2006 Igor Pavlov
#9 33.73 Copyright (C) 2016-2023 Viktor Kirilov
#9 33.73 UPX comes with ABSOLUTELY NO WARRANTY; for details type 'upx_devel4_packed -L'.
#9 DONE 33.9s
```

Maybe you fix it, I will test again if you release `v4.3.0`.

**Comment by tosone** [comment]
You can close this issue now. ^-^
